File: java_template/src/main/java/lambda/Transform.java
package lambda;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.S3Object;
import com.amazonaws.services.s3.model.GetObjectRequest;
import saaf.Inspector;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.text.DecimalFormat;
import org.apache.commons.csv.CSVParser;
import org.apache.commons.csv.CSVRecord;
import org.apache.commons.csv.CSVFormat;


public class Transform implements RequestHandler<HashMap<String, Object>, HashMap<String, Object>> {

    public CSVParser readCSV(
        Inspector inspector, 
        AmazonS3 s3Client, 
        String bucketName, 
        String fileName
    ) {
        S3Object s3Object = s3Client.getObject(new GetObjectRequest(bucketName, fileName));
        InputStream csvInputStream = s3Object.getObjectContent();
        InputStreamReader csvInputStreamReader = new InputStreamReader(csvInputStream);
        CSVParser csvParser = null;
        try {
            csvParser = CSVFormat.EXCEL.withHeader().parse(csvInputStreamReader);
        } catch (IOException e) {
            inspector.addAttribute("error", "Error while reading CSV file from S3");
            inspector.addAttribute("stackTrace", e.getStackTrace());
        }

        return csvParser;
    }

    public String processCSV(CSVParser csvParser) {
        Iterator<CSVRecord> csvIterator = csvParser.iterator();
        ArrayList<String> uniqueOrderIDList = new ArrayList<String>();
        StringBuilder stringBuilder = new StringBuilder();

        // Create header
        ArrayList<String> header = new ArrayList<String>();
        header.add("Region");
        header.add("Country");
        header.add("Item Type");
        header.add("Sales Channel");
        header.add("Order Priority");
        header.add("Order Date");
        header.add("Order ID");
        header.add("Ship Date");
        header.add("Units Sold");
        header.add("Unit Price");
        header.add("Unit Cost");
        header.add("Total Revenue");
        header.add("Total Cost");
        header.add("Total Profit");
        header.add("Order Processing Time");
        header.add("Gross Margin");

        stringBuilder.append(String.join(",", header) + "\n");

        // Process fields
        while (csvIterator.hasNext()) {
            CSVRecord csvRecord = csvIterator.next();

            // Remove duplicated order IDs
            String orderID = csvRecord.get("Order ID");
            if (!uniqueOrderIDList.contains(orderID)) {
                uniqueOrderIDList.add(orderID);
                ArrayList<String> row = new ArrayList<String>();

                // Add fields
                for (String key : header) {
                    if (key.equals("Order Priority")) { // Correct order priority
                        String orderPriority = csvRecord.get("Order Priority"); // Add Order Priority
                        if (orderPriority.equals("L")) {
                            row.add("Low");
                        }
                        else if (orderPriority.equals("M")) {
                            row.add("Medium");
                        }
                        else if (orderPriority.equals("C")) {
                            row.add("Critical");
                        }
                        else if (orderPriority.equals("H")) {
                            row.add("High");
                        } else {
                            row.add("Unknown");
                        }
                    } else if (key.equals("Order Processing Time")) { // Add "Order Processing Time"
                        String[] orderDate = csvRecord.get("Order Date").split("/");
                        String[] shipDate = csvRecord.get("Ship Date").split("/");

                        Integer orderProcessingTime = (Integer.parseInt(shipDate[2]) - Integer.parseInt(orderDate[2]))*365 + 
                            (Integer.parseInt(shipDate[0]) - Integer.parseInt(orderDate[0]))*30 + 
                                (Integer.parseInt(shipDate[1]) - Integer.parseInt(orderDate[1]));
                        row.add(orderProcessingTime.toString());
                    } else if (key.equals("Gross Margin")) { // Add Gross Margin
                        Float grossMarginFloat = Float.parseFloat(csvRecord.get("Total Profit")) / Float.parseFloat(csvRecord.get("Total Revenue"));
                        DecimalFormat decimalFormat = new DecimalFormat("0.00");
                        String grossMargin = decimalFormat.format(grossMarginFloat);
                        row.add(grossMargin);
                    } else {
                        row.add(csvRecord.get(key));
                    }
                }

                stringBuilder.append(String.join(",", row) + "\n");
            }
        }

        return stringBuilder.toString();
    }

    public void writeCSV(
        AmazonS3 s3Client, 
        String bucketName, 
        String fileName, 
        String csvString
    ) {
        byte[] csvBytes = csvString.getBytes();
        InputStream csvInputStream = new ByteArrayInputStream(csvBytes);
        ObjectMetadata meta = new ObjectMetadata();
        meta.setContentLength(csvBytes.length);
        meta.setContentType("text/plain");

        s3Client.putObject(bucketName, fileName, csvInputStream, meta);
    }

    public HashMap<String, Object> handleRequest(HashMap<String, Object> request, Context context) {
        
        Inspector inspector = new Inspector();

        AmazonS3 s3Client = AmazonS3ClientBuilder.standard().build();
        String bucketName = (String) request.get("bucketName");
        String fileName = (String) request.get("fileName");

        CSVParser csvParser = readCSV(inspector, s3Client, bucketName, fileName);
        String csvString = processCSV(csvParser);
        writeCSV(s3Client, bucketName, "Processed_" + fileName, csvString);

        return inspector.finish();
    }
}

File: python_template/src/transform.py
import os
import sys
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__))))

import json
import logging
from Inspector import *
import time
import boto3
import numpy as np
import pandas as pd
import json
from io import StringIO


def transform(csv):
    df = csv
    df = df.drop_duplicates(subset=['Order ID'], keep='first')   #Remove duplicate data identified by [Order ID]

    # Transform [Order Priority] column
    col_no = [i for i in df.columns].index('Order Priority') 
    for i in range(len(df)):
        if df.iloc[i, col_no] == "L":
            df.iloc[i, col_no] = "Low"
        elif df.iloc[i, col_no] == "M":
            df.iloc[i, col_no] = "Medium"
        elif df.iloc[i, col_no] == "H":
            df.iloc[i, col_no] = "High"
        elif df.iloc[i, col_no] == "C":
            df.iloc[i, col_no] = "Critical"

    # Add a [Gross Margin] column
    Gross_Margin = df['Total Profit'].values / df['Total Revenue'].values
    df['Gross Margin'] = Gross_Margin

    # Add column [Order Processing Time]
    Ship_Date = df['Ship Date'].values.tolist()
    for i in range(len(Ship_Date)):
        Ship_Date[i] = Ship_Date[i].split('/')
        Ship_Date[i] = list(map(int, Ship_Date[i]))

    Order_Date = df['Order Date'].values.tolist()
    for i in range(len(Order_Date)):
        Order_Date[i] = Order_Date[i].split('/')
        Order_Date[i] = list(map(int, Order_Date[i]))

    Order_Processing_Time = [0 for _ in range(len(df))]
    for i in range(len(df)):
        Order_Processing_Time[i] = (Ship_Date[i][2] - Order_Date[i][2]) * 365 + (Ship_Date[i][0] - Order_Date[i][0]) * 30 + (Ship_Date[i][1] - Order_Date[i][1])

    #write CSV
    df['Order Processing Time'] = Order_Processing_Time

    return df

def handler(request, context):
    # Import the module and collect data
    inspector = Inspector()

    key = str(request['key'])
    bucketname = str(request['bucketname'])
    
    # read csv from s3
    s3 = boto3.client('s3')
    csv_file = s3.get_object(Bucket=bucketname, Key=key)
    csv = pd.read_csv(csv_file['Body'])
    # transform csv
    csv_new = transform(csv)
    csv_buffer = StringIO()
    csv_new.to_csv(csv_buffer, index=False)
    
    content = csv_buffer.getvalue()
    # write csv to s3
    s3.put_object(Bucket=bucketname, Key="Processed_{}".format(key), Body=content)

    return inspector.finish()


if __name__ == "__main__":
    context = None
    request = {}
    request["key"] = "100_Sales_Records.csv"
    request["bucketname"] = "test.bucket.562f20.leb"

    response = handler(request, context)
    print(response)
